Use ORC JIT for better compile time and runtime performance

Keno opened the pull request:
This is somewhat of a rewrite of the way we use MCJIT + a move to ORC JIT to avoid a) some runtime overhead related to MCJIT not reusing the pass manager and b) the memory overhead of MCJIT requiring you to hold onto the LLVM IR, both of which will not be fixed upstream, because they are fixed in the C API. The new ORC based jit is in jitlayers.cpp, but the intention is that both the old jit and MCJIT continue working for now. As a result the code is slightly more messy than it should be in a number places because there are now three possible JITs that could run over it.

There are two core aspects of this rewrite:
1) Attempting to combine multiple functions into a single module in order to reduce per-module overhead in the compiler. The primary difficulty here is that sometimes we need to compile functions while we're in the middle of IRgening another function (most notably in inference and in staged functions). As a result, we cannot simply have just one module into which everything gets emitted, because that way LLVM would attempt to generate code for a partially IRgened function, which is not what we want. Instead what this IR does is to temporarily put all functions into a global module (technically they could be parentless, but some of LLVM's APIs don't like dealing with parentless functions), which then get reparented to a single active module as soon as all function in a given call cycle are complete.
2) Being more aggressive about freeing memory at all stages of using LLVM. This includes freeing the IR module, as soon as the native code is generated and freeing the native code as soon as it's moved to the executable pages (both of which we were holding on to). The one caveat to this is that there is an additional copy of the native code for the debugger (for the debug sections, in the future I will make sure that copy does not include the actual generated code). Our backtracing code now shares that copy for reading DWARF information.

I have two more patches pending upstream (1 accepted but not landed, one pending review) before this will actually work with an unmodified copy of LLVM, but I expect to be able to land them shortly. Once that's done, I'll create a backported version for LLVM 3.7.1, which we can then officially activate in a separate pull request.

As I mentioned, this should run on all three JITs, so the transition should be relatively smooth. The idea is to run on ORC JIT for LLVM 3.8 and LLVM 3.7.1 with patches, MCJIT on LLVM 3.5+ without patches (this will have some of the known performance problems) where distributions refuse to ship our patches and the old JIT on LLVM 3.3. I'm anticipating keeping the 3.3 JIT around for at least a little while longer in case we encounter any major issues (e.g. in packages) after switching people over, but hopefully we'll be able to remove support for it soon. I imagine the most like scenario is that we wait for the release of 0.5/LLVM 3.8 and go full ORC jit on master, at which point we'll be able to delete quite a bit of code.

Keno committed:
Rewrite JIT based on ORC

This is somewhat of a rewrite of the way we use MCJIT + a move to ORC JIT to avoid a) some runtime overhead related to MCJIT not reusing the pass manager and b) the memory overhead of MCJIT requiring you to hold onto the LLVM IR, both of which will not be fixed upstream, because they are fixed in the C API. The new ORC based jit is in jitlayers.cpp, but the intention is that both the old jit and MCJIT continue working for now. As a result the code is slightly more messy than it should be in a number places because there are now three possible JITs that could run over it.

There are two core aspects of this rewrite:
1) Attempting to combine multiple functions into a single module in order to reduce per-module overhead in the compiler. The primary difficulty here is that sometimes we need to compile functions while we're in the middle of IRgening another function (most notably in inference and in staged functions). As a result, we cannot simply have just one module into which everything gets emitted, because that way LLVM would attempt to generate code for a partially IRgened function, which is not what we want. Instead what this IR does is to temporarily put all functions into a global module (technically they could be parentless, but some of LLVM's APIs don't like dealing with parentless functions), which then get reparented to a single active module as soon as all function in a given call cycle are complete.
2) Being more aggressive about freeing memory at all stages of using LLVM. This includes freeing the IR module, as soon as the native code is generated and freeing the native code as soon as it's moved to the executable pages (both of which we were holding on to). The one caveat to this is that there is an additional copy of the native code for the debugger (for the debug sections, in the future I will make sure that copy does not include the actual generated code). Our backtracing code now shares that copy for reading DWARF information.

Keno commented:
Also, this probably needs some cleanup.

tkelman commented on src/cgutils.cpp:
```diff
@@ -27,48 +29,183 @@ static Instruction *tbaa_decorate(MDNode* md, Instruction* load_or_store)
+static GlobalVariable *CreateGlobalString(StringRef Str,
+                                                  const Twine &Name,
+                                                  unsigned AddressSpace) {
+  Constant *StrConstant = ConstantDataArray::getString(jl_LLVMContext, Str);
```
indentation

tkelman commented on src/jitlayers.cpp:
```diff
@@ -0,0 +1,392 @@
+// This file is part of Julia.
+// Parts of this file are copied from LLVM, under the UIUC license.
```
add this to the lists in `LICENSE.md` and `contrib/add_license_to_files.jl`

vtjnash replied:
ianal, but i this doesn't sound like a valid license, and it doesn't sound like a good situation for this file. i would prefer that any content you add that requires a special license to be placed in a separate file.

fwiw, i used Kaleidoscope tutorial as the reference for learning how to write an OrcJIT, but the actual implementation of JuliaOJIT I believe should be considered as MIT licensed like the rest of codegen

Keno replied:
This was meant to be addressed by the ```// ------------------------ TEMPORARILY COPIED FROM LLVM -----------------``` ```//-------------------``` markers.

tkelman commented on test/llvmcall.jl:
```diff
@@ -74,15 +74,15 @@ end
+#    llvmcall("""%2 = call double @llvm.ceil.f64(double %0)
+#        ret double %2""", Float64, Tuple{Float64}, x)
+#end
+#@test_throws ErrorException undeclared_ceil(4.2)
```
what was the issue with this test?

Keno replied:
It only accidentally worked on llvm 3.5, because we were using a module per function call. On this branch, everything gets parsed into a global module, so if ceil was previously declared this does not actually throw an error. A lot of this is really a limitation of the LLVM IR parsing API, so I'm planning to redo this properly, potentially with some LLVM API changes, at which point there will be more logical and complete error messages here, but that is out of the scope of this PR.

tkelman commented on src/gf.c:
```diff
@@ -1457,7 +1457,7 @@ JL_DLLEXPORT jl_value_t *jl_matching_methods(jl_function_t *gf,
 
 // compile-time method lookup
-jl_function_t *jl_get_specialization(jl_function_t *f, jl_tupletype_t *types)
+jl_function_t *jl_get_specialization(jl_function_t *f, jl_tupletype_t *types, void *cyclectx)
```
there's a use of this in `threading.c` that needs updating

tkelman replied:
bump, still needed to not break threading build

vtjnash replied:
bump again.

Keno commented:
After the rebase, there's a heisenbug on OS X when using this branch. Posting here in case something jumps out to anybody:
```
frame #2: 0x0000000100025b74 libjulia-debug.dylib`jl_lam_vinfo(l=0x000000010a1a85f0) + 212 at ast.c:706
   703 	{
   704 	    assert(jl_is_expr(l));
   705 	    jl_value_t *le = jl_exprarg(l, 1);
-> 706 	    assert(jl_is_array(le));
   707 	    jl_value_t *ll = jl_cellref(le, 0);
   708 	    assert(jl_is_array(ll));
   709 	    return (jl_array_t*)ll;
(lldb) p le
(jl_value_t *) $1 = 0x0000000108022470
(lldb) p jl_(le)
22
(lldb) p l
(jl_expr_t *) $2 = 0x000000010a1a85f0
(lldb) p jl_(l)
Expr(:lambda, <?#0x109d31e10::TypeName(name=:DataType, module=Core, names=svec(:name, :super, :parameters, :types, :instance, :size, :abstract, :mutable, :pointerfree, :ninitialized), primary=DataType, cache=svec(), linearcache=svec(), uid=3)>, 22, TypeName(name=:NTuple, module=Core, names=svec(), primary=NTuple, cache=svec(), linearcache=svec(), uid=20))::Any
```

yuyichao commented:
From your last line, looks like the object (and tag) is corrupted?

Keno commented:
Sure does! If this were Linux I'd rr it and find out, but it's not. Just wondering if there was something within the last week or so that might have caused this.

yuyichao commented:
Just had a quick look at all the assignments to `->ast`. Should [this](https://github.com/JuliaLang/julia/blob/cb52f15feec9052e68c9102e7e12e93a49367def/src/gf.c#L1580) be `jl_gc_wb(unspec->linfo, unspec->linfo->ast)` instead?

This is indeed new from https://github.com/JuliaLang/julia/commit/15a4a21b309d01f7bc68bf1f96b3fcea26e3fddc

Keno commented:
@yuyichao You're a saint! That indeed seems to have been the culprit (or at least it compiled successfully for the first time today after making that change)!

Keno commented:
Unfortunately, I may have spoken too soon (though the symptom changed, so maybe it's just a new problem).

Keno commented:
I think you definitely found a real issue though. I'm having trouble reproducing the new crash.

yuyichao commented:
What does the new crash look like?

Keno commented:
```
version.jl
LLVM ERROR: Program used external function '__j_str96�uo�����thod#951' which could not be resolved!
*** This error is usually fixed by running `make clean`. If the error persists, try `make cleanall`. ***
```
```
exports.jl
essentials.jl
docs/bootstrap.jl
base.jl
reflection.jl
options.jl
promotion.jl
tuple.jl
julia(14187,0x7fff7348a000) malloc: *** error for object 0x7ffe7522fae8: incorrect checksum for freed object - object was probably modified after being freed.
```

Keno commented:
The new crash may be specific to MEMDEBUG mode, which I had enabled to track down the other issue. Let's see.

yuyichao commented:
There might be a few other places that are technically missing write barriers. I seriously doubt if any of them can actually trigger though......

```
diff --git a/src/ast.c b/src/ast.c
index 02f548d..0b00d6e 100644
--- a/src/ast.c
+++ b/src/ast.c
@@ -377,8 +377,10 @@ static jl_value_t *scm_to_julia_(value_t e, int eo)
             }
             jl_expr_t *ex = jl_exprn(sym, n);
             // allocate a fresh args array for empty exprs passed to macros
-            if (eo && n == 0)
+            if (eo && n == 0) {
                 ex->args = jl_alloc_cell_1d(0);
+                jl_gc_wb(ex, ex->args);
+            }
             for(i=0; i < n; i++) {
                 assert(iscons(e));
                 jl_cellset(ex->args, i, scm_to_julia_(car_(e),eo));
diff --git a/src/gf.c b/src/gf.c
index ee49f0b..27b5e33 100644
--- a/src/gf.c
+++ b/src/gf.c
@@ -983,6 +983,7 @@ JL_DLLEXPORT jl_function_t *jl_instantiate_staged(jl_methlist_t *m,
         ex = oldast;
     }
     func = (jl_function_t*)jl_toplevel_eval_in_warn(m->func->linfo->module, (jl_value_t*)ex, 1); // need to eval macros in the right module, but not give a warning for the `eval` call unless that results in a call to `eval`
+    // no WB for symbols
     func->linfo->name = m->func->linfo->name;
     JL_GC_POP();
     return func;
@@ -2075,6 +2076,7 @@ void jl_add_method(jl_function_t *gf, jl_tupletype_t *types, jl_function_t *meth
             // already used by another GF; make a copy (issue #10373)
             meth = jl_instantiate_method(meth, jl_emptysvec, NULL);
         }
+        // no WB for symbols
         meth->linfo->name = n;
     }
     if (isstaged && tvars != jl_emptysvec) {
diff --git a/src/interpreter.c b/src/interpreter.c
index f7b3010..ed1dcb0 100644
--- a/src/interpreter.c
+++ b/src/interpreter.c
@@ -356,6 +356,7 @@ static jl_value_t *eval(jl_value_t *e, jl_value_t **locals, size_t nl, size_t ng
         super = eval(args[2], locals, nl, ngensym);
         jl_set_datatype_super(dt, super);
         b->value = temp;
+        jl_gc_wb_binding(b, temp);
         if (temp==NULL || !equiv_type(dt, (jl_datatype_t*)temp)) {
             jl_checked_assignment(b, (jl_value_t*)dt);
         }
@@ -387,6 +388,7 @@ static jl_value_t *eval(jl_value_t *e, jl_value_t **locals, size_t nl, size_t ng
         super = eval(args[3], locals, nl, ngensym);
         jl_set_datatype_super(dt, super);
         b->value = temp;
+        jl_gc_wb_binding(b, temp);
         if (temp==NULL || !equiv_type(dt, (jl_datatype_t*)temp)) {
             jl_checked_assignment(b, (jl_value_t*)dt);
         }
@@ -432,6 +434,7 @@ static jl_value_t *eval(jl_value_t *e, jl_value_t **locals, size_t nl, size_t ng
         }
         JL_CATCH {
             b->value = temp;
+            jl_gc_wb_binding(b, temp);
             jl_rethrow();
         }
         for(size_t i=0; i < jl_svec_len(para); i++) {
@@ -444,6 +447,7 @@ static jl_value_t *eval(jl_value_t *e, jl_value_t **locals, size_t nl, size_t ng
         }
 
         b->value = temp;
+        jl_gc_wb_binding(b, temp);
         if (temp==NULL || !equiv_type(dt, (jl_datatype_t*)temp)) {
             jl_checked_assignment(b, (jl_value_t*)dt);
         }
```

yuyichao commented:
Actually isn't the `ex->args = jl_alloc_cell_1d(0);` one missing a GC root for `ex`?

Keno commented:
Another symptom of the crash:
```
GC error (probable corruption) :
<?#0x112a16fb0::0x0>
```
Current investigation suggested only MEMDEBUG, julia debug and OS X.

Keno commented:
@yuyichao gc is disabled in `scm_to_julia_`

yuyichao commented:
Ahh, true, I'm always confused since that function **does** have gc frames.....

Have you tried to launch the process with ASLR off so that you can get a repeatable corruption address? http://stackoverflow.com/questions/23897963/documented-way-to-disable-aslr-on-os-x

Keno commented:
Might still need a write barrier though, even if gc is disabled? I think I was able to catch the corruption using guard malloc, so hold on.

yuyichao commented:
I think the write barrier technically not needed since with GC disabled, both objects are newly allocated and are guaranteed to be both young. I'd still prefer to include it just for good measure but it shouldn't cause the corruption you see.

Keno commented:
I think I found it and it was in one of my LLVM patches.

Keno committed:
Don't insit on inlining Function* llvmcalls

Since they can drag with them a significant number of other functions,
it is easier to just call them. In the future, we'll probably want to
run an LLVM inling pass, which will restore any benefit this had for
performance. Still keep the inlining for the text version of llvmcall
though in case people rely on it.

tkelman commented on src/cgutils.cpp:
```diff
@@ -230,43 +420,17 @@ class FunctionMover : public ValueMaterializer
+            destModule->getFunctionList().push_back(NewF);
         }
-	return NewF;
+	    return NewF;
```
replace the tab with spaces here

Keno committed:
Make sure `other` is not referenced anywhere in jl_get_llvmf

When regenerating the LLVM IR, make the jlcall wrapper use the declaration
pointing to the original function rather than the newly generated function.
That way it is safe to delete `other` in jl_get_llvmf. In the future it
would be good to have a function that just recompiles the wrapper without
always compiling both, but that is out of the scope of this PR.

Keno commented:
Squashed + Rebased. @vtjnash wanted to take one more look, but I think this is ready to go after.

tkelman commented on src/cgutils.cpp:
```diff
@@ -27,47 +29,181 @@ static Instruction *tbaa_decorate(MDNode* md, Instruction* load_or_store)
+    return GV;
+}
+
+/// \brief Same as CreateGlobalString, but return a pointer with "i8*" type
```
doxygen?

vtjnash replied:
it seems there should also be a comment about why we aren't able to use the IRBuilder functions for these

tkelman commented on src/cgutils.cpp:
```diff
@@ -27,47 +29,181 @@ static Instruction *tbaa_decorate(MDNode* md, Instruction* load_or_store)
+        Value::use_iterator UI2 = Expr->use_begin(), E2 = Expr->use_end();
+        for (; UI2 != E2;) {
+            Use &Use2 = *UI2;
+             ++UI2;
```
indent

vtjnash commented on src/intrinsics.cpp:
```diff
@@ -1157,10 +1151,10 @@ static Value *emit_untyped_intrinsic(intrinsic f, Value *x, Value *y, Value *z,
-      return builder.CreateCall(fmaintr,{ FP(x), FP(y), FP(z) });
+      return builder.CreateCall(prepare_call(fmaintr),{ FP(x), FP(y), FP(z) });
 #else
       return builder.CreateCall3(fmaintr, FP(x), FP(y), FP(z));
```
`prepare_call`? or do distros not care about llvm34-36 anymore

Keno replied:
Can't hurt.

vtjnash commented on src/jitlayers.cpp:
this could just be a warning

vtjnash commented on src/jitlayers.cpp:
missing { } braces here (Julia-style not llvm style :)

vtjnash commented on src/jitlayers.cpp:
i'm not sure if i understand clearly why this is unfortunate, but since it is currently in a `vector<unique_ptr<...>>` that will be destroyed as soon as this function returns, taking ownership would presumably mean calling `Object.release()` here

vtjnash commented on src/jitlayers.cpp:
i vaguely feel like you meant `SavedObjects` to be `std::vector<std::unique_ptr<OwningBinary<ObjectFile>>>` so that it was managing the object lifetime (although we don't actually implement deletion at this time)

vtjnash commented on src/jitlayers.cpp:
`PM.run` reuse is known to be broken on LLVM 3.8.1. for distros, we should probably handle that case

vtjnash commented on src/jitlayers.cpp:
this todo was from the tutorial toy implementation. we should perhaps print an error and the IR, then abort?

vtjnash commented on src/jitlayers.cpp:
i think this PR subsumes my comment here, with its implementation of module coalescing

vtjnash commented on src/jitlayers.cpp:
in the FunctionMover, I thought you indicated this search may be expensive and `O(n)` in the number of modules that have been emitted. If so, and if you're still looking for places where performance is potentially being lost, this may be one place to consider.

vtjnash commented on src/jitlayers.cpp:
is it worth returning the stub from `active_module`? it looks like the FunctionMover has a quasi-dependency on this

Keno commented on src/jitlayers.cpp:
The plan is to not use ORC on 3.7.1 (without patches) and continue to use MCJIT there.

Keno commented on src/jitlayers.cpp:
ok

Keno commented on src/jitlayers.cpp:
The memory buffer itself is owned by https://github.com/llvm-mirror/llvm/blob/813f44a29fd0fd140127023222d0633e23783bcc/include/llvm/ExecutionEngine/Orc/IRCompileLayer.h#L70, for which we do not have a reference.

Keno commented on src/jitlayers.cpp:
I was under the impression that OwningBinary already manages the lifetime. Why would the extra indirection through unique_ptr help?

Keno commented on src/jitlayers.cpp:
Perhaps, but it works fine for now, so I'll reserve potential performance improvements for post-merge with careful profiling.

Keno commented on src/jitlayers.cpp:
Actually, wouldn't printing a warning most likely allocate, defeating the purpose of this?

Keno commented on src/jitlayers.cpp:
I think it might still be reasonable to change the way we do bootstrapping once we get rid of MCJIT

Keno commented on src/jitlayers.cpp:
Yes, I think that's not a bad idea.

Keno commented on src/jitlayers.cpp:
Actually, probably just fix this in the FunctionMover directly, since this has little to do with the ExecutionEngine

Keno commented on src/jitlayers.cpp:
Thinking again, I don't think there's actually any case any more where we need to look in any module other than the source module (during bootstrap that will always be the shadow module), so probably just ifdef out this check in ORCJIT mode.

Keno committed:
Address review comments

vtjnash commented on src/jitlayers.cpp:
possibly, but that's the problem with virtual address space limits (`ulimit -v`)

Keno commented on src/jitlayers.cpp:
Fair point.

vtjnash commented on src/jitlayers.cpp:
ah, good point (although, to be pedantic, at this point in the code, it is actually owned by https://github.com/llvm-mirror/llvm/blob/813f44a29fd0fd140127023222d0633e23783bcc/include/llvm/ExecutionEngine/Orc/IRCompileLayer.h#L66, for which we also don't have a reference)

vtjnash commented on src/jitlayers.cpp:
oops, you're right, it wouldn't. i was thinking OwningBinary was a pointer and forgetting that it is already a value type.

Keno committed:
Turn allocation failure into a warning

Keno commented:
I'm gonna go ahead and merge this. I think anything else can be addressed post-merge or by separate cleanup later.
